lib/class-wp-style-engine.php

```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * WP_Style_Engine class
+ *
+ * @package Gutenberg
+
+ */
+
+/**
+ * Singleton class representing the style engine.
+ *
+ * Consolidates rendering block styles to reduce duplication and streamline
+ * CSS styles generation.
+ *
+ * @since 6.0.0
+ */
+class WP_Style_Engine_Gutenberg {
+	/**
+	 * Registered CSS styles.
+	 *
+	 * @since 5.5.0
+	 * @var array
+	 */
+	private $registered_styles = array();
+
+	/**
+	 * Container for the main instance of the class.
+	 *
+	 * @since 5.5.0
+	 * @var WP_Style_Engine_Gutenberg|null
+	 */
+	private static $instance = null;
+
+	public function __construct() {
+		// Borrows the logic from `gutenberg_enqueue_block_support_styles`.
+		$action_hook_name = 'wp_footer';
+		if ( wp_is_block_theme() ) {
+			$action_hook_name = 'wp_enqueue_scripts';
+		}
+		add_action(
+			$action_hook_name,
+			array( $this, 'output_styles' )
+		);
+	}
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Assemble the style rule from a list of rules, and store based on a key
+	 * generated from the class name, the selector, and any values used as a suffix.
+	 *
+	 * @param string $key     A class name used to construct a key.
+	 * @param array  $options An array of options, rules, and selector for constructing the rules.
+	 *
+	 * @return string The class name for the added style.
+	 */
+	public function add_style( $key, $options ) {
+		$class    = ! empty( $options['suffix'] ) ? $key . '-' . sanitize_title( $options['suffix'] ) : $key;
```

Comment:
Too early to think about BEM? E.g., `wp-layout__flex-orientation-vertical--justify-right`  😄

Reply:
I think a BEM-like naming convention would be a good way to go if we wind up using this highly-specific class based approach 👍

Reply:
I know it's implied – and also that this is all highly experimental – but I guess we'll have to replicate all this in the editor(s) as well.

Reply:
Yeah, totally. I figured a good place to start is with support for the front-end of the site, and if we settle on an approach, then we'd want to roll it out everywhere. I'm imagining this PR more as a discussion point rather than necessarily the right path forward, but happy to keep hacking away here, too, of course!